### PR TITLE
IngestionService integration test

### PR DESCRIPTION
**Pull request by andreidan**

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            this.expressionAnalysisContext = new ExpressionAnalysisContext();
+            this.expressionAnalyzer = new ExpressionAnalyzer(functions, null, null, inputColumnProvider, null);
+        }
+
+        void registerIngestionImplementation() {
+            ingestionService.registerImplementation(INGESTION_SOURCE_ID, this);
+        }
+
+        void ingestData(String topic, int data) {
+            if (ingestRules.size() > 0) {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: this is not needed

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                for (IngestRule ingestRule : ingestRules) {
+                    Symbol conditionSymbol = expressionAnalyzer.convert(SqlParser.createExpression(ingestRule.getCondition()), expressionAnalysisContext);
+                    Predicate<Row> conditionPredicate = RowFilter.create(inputFactory, conditionSymbol);
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: both lines above should be done while applying rules, at least on real implementations. not sure if we don't care about this here, I fear that someone is simply copying this pattern.

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                    if (conditionPredicate.test(new Row1(topic))) {
+                        execute(String.format(Locale.ENGLISH,
+                            "insert into %s values (%s);", ingestRule.getTargetTable(), data)
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: `data` should never be passed in using string format. even that this is a test, I wouldn't do it ;)

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    private void resumeDataIngestion() {
+        dataFlowSemaphore.release();
+    }
+
+    @Before
+    public void setupIngestRuleAndImplementation() {
+        execute("create table ingest_data_raw (data int)");
+        execute(String.format(Locale.ENGLISH,
+            "create ingest rule %s on %s where topic = '%s' into %s",
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: would prefer using parameterSubstitution wherever possible (avoiding people would copy this pattern ;)

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        }, "DataIngestionThread");
+        dataIngestionThread.start();
+    }
+
+    @After
+    public void dropIngestRuleAndStopProducingData() throws InterruptedException {
+        execute("drop ingest rule if exists " + UNDER_TEST_INGEST_RULE_NAME);
+        resumeDataIngestion();
+        produceData = false;
+        dataIngestionThread.join();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: pls add a timeout

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            }
+        }, "DataIngestionThread");
+        dataIngestionThread.start();
+    }
+
+    @After
+    public void dropIngestRuleAndStopProducingData() throws InterruptedException {
+        execute("drop ingest rule if exists " + UNDER_TEST_INGEST_RULE_NAME);
+        resumeDataIngestion();
+        produceData = false;
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: shouldn't this be done first before resuming?

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    @Test
+    public void testDropIngestRuleWillCauseImplementationToStopIngesting() throws Exception {
+        implementation.registerIngestionImplementation();
+        dataFlowingLatch.await(5, TimeUnit.SECONDS);
+
+        pauseDataIngestion();
+        execute("drop ingest rule " + UNDER_TEST_INGEST_RULE_NAME);
+        int expectedLastIngestedValue = lastProducedData.get();
+        resumeDataIngestion();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: afaik ingestion must be paused to get the accurate last produced/ingested data. can you pls add a comment here?

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                    IngestRule ingestRule = predicateAndRuleTuple.v2();
+                    execute("insert into " + ingestRule.getTargetTable() + " values (?);", new Object[]{data});
+                    execute("refresh table " + ingestRule.getTargetTable());
+                }
+            }
+        }
+
+        @Override
+        public void applyRules(Set<IngestRule> rules) {
+            this.conditionPredicateAndIngestRules = Collections.synchronizedSet(new HashSet<>(rules.size()));
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: using an AtomicReference for the instance var and just apply it once the new set is build would be better

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            }
+        };
+
+        TestIngestionImplementation(Functions functions,
+                                    IngestionService ingestionService) {
+            this.conditionPredicateAndIngestRules = Collections.synchronizedSet(new HashSet<>());
+            this.inputFactory = new InputFactory(functions);
+            this.ingestionService = ingestionService;
+            this.expressionAnalysisContext = new ExpressionAnalysisContext();
+            Row args = new RowN($(DATA_TOPIC));
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: this wouldn't work for a real implementation

File: sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        assertThat(response.rows()[1][0], is(lastProducedData.get()));
+    }
+
+    @Test
+    public void testDropIngestRuleWillCauseImplementationToStopIngesting() throws Exception {
+        dataFlowingLatch.await(5, TimeUnit.SECONDS);
+        pauseDataIngestion();
+        execute("drop ingest rule " + UNDER_TEST_INGEST_RULE_NAME);
+        int expectedLastIngestedValue = lastProducedData.get();
+        // data that's flowing in after the rule was dropped should not be inserted into the target table
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: the explanation at the comment would not require a pause/resume, data should not be ingested when a rule is dropped anyway. afaik the point here is just about resolving a consistent lastProducedData to get valid test results

Comment (andreidan) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: no, as the test name suggests (does it need a better name ?) we are testing that the implementation doesn't insert anything if it has no rules.
"data should not be ingested when a rule is dropped anyway" YES, that's exactly what this is testing

Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: my point is: why do we need to pause/resume? this does imply somehow that the implementation is not stopping to ingest when the rule is dropped...

Comment (andreidan) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: pause to get the expected value. without that what can we assert on? 
resume having data flow in when we have no rules.

"this does imply somehow that the implementation is not stopping to ingest when the rule is dropped". That's exactly what we are asserting here, that the implementation stopped ingesting - hence assert the last value inserted was the one before the rule was dropped

Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/IngestionServiceIntegrationTest.java]: can you pls just document why you are pausing the ingestion?

**Pull request merged**